src/main/java/uk/gov/gchq/palisade/util/UriBuilder.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2020 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.gchq.palisade.util;
+
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Optional;
+
+public class UriBuilder {
+
+    private Optional<URI> baseUri = Optional.empty();
+
+    public static class AuthorityBuilder {
+        private Optional<URI> baseUri = Optional.empty();
+
+        public static IUserInfo create(final URI baseUri) {
+            AuthorityBuilder builder = new AuthorityBuilder();
+            builder.baseUri = Optional.of(baseUri);
+            return builder.build();
+        }
+
+        public static IUserInfo create() {
+            return new AuthorityBuilder().build();
+        }
+
+        private IUserInfo build() {
+            return userInfo -> host -> port -> String.format(
+                    "//%s%s%s",
+                    Optional.ofNullable(userInfo).or(() -> baseUri.map(URI::getUserInfo)).map(str -> str + "@").orElse(null),
+                    Optional.ofNullable(host).or(() -> baseUri.map(URI::getHost)).orElseThrow(),
+                    Optional.ofNullable(port).or(() -> baseUri.map(URI::getPort)).map(str -> ":" + port).orElse(null)
+            );
+        }
+
+        public interface IUserInfo {

```

Comment:
Add JavaDoc on these public interface classes. In particular I don't understand the need for the UserInfo in a URI builder.

Reply:
Included for completeness, to include an authority component

Reply:
Each uri may optionally have an authority, which itself may optionally have UserInfo, Host and Port

Reply:
Added short javadoc